### DISCUSSION TITLE
deploying with reverse proxy under root/app folder

### DISCUSSION THREAD

**Discussion by dncpax** [lowdefy/lowdefy#719] [Questions and Help] [answered]
Hi there. Been looking at lowdefy and how it works... so bare with me if these are basic questions...

My question is how to publish an app from a www server with a reverse proxy to lowdefy app, but not take a whole domain, and instead tuck it under a sub-folder, like http://host/app1, /app2 and so on.

As I understand it you can run lowdefy app1 as an Express app, and have a specific port for app1, like host:xxx.
So the reverse proxy would proxy all requests that hit http://host/app1/* to host:xxx/*.

Is this kind of setup possible at all?

Thanks.

**Comment by Gervwyk** [comment, marked as answer]
Hi @dncpax 
No worries, all questions are welcome :)
So unfortunately this is not currently possible. We'll investigate to support this in the next major release. This has come up before, I'll create a feature request issue for this #720

**Reply by dncpax** [reply]
Thanks for replying so quickly. That's unlucky of me...

First I was expecting to build and xcopy over to a server, like discussed in #537 ... but that's not possible either. So turned to reverse proxying.

Ok, I guess I'll wait to see what happens next.

Thanks again.

**Reply by Radiergummi** [reply]
@dncpax couldn't you place nginx in front of Lowdefy, and have it rewrite the URIs before passing requests to Lowdefy, and substitute it in the response body? Something like this:

```nginx
location /app1/ {
    proxy_pass http://localhost:8080/;
    sub_filter "http://your.server.example/" "http://your.server.example/app1/";
    sub_filter_once off;
}
```

**Reply by dncpax** [reply]
@Radiergummi yes I have redash running like that. But it's a struggle every new release brings new strings you have to find a way to replace and it gets complicated to mantain. Sometimes it just doesn't work. So it's a risk going that way.

**Reply by dncpax** [reply]
on the other hand you have Jenkins that's a breeze to setup with a reverse proxy and under a path.

**Reply by SamTolmay** [reply]
We haven't worked with nginx or reverse proxies before, but it would be great for us if this could be supported and work well. If there is anything you need, maybe a configuration option we should expose or something like that, please create a feature request for this. We will also be happy to accept any contributions with regards to this.

**Reply by dncpax** [reply]
it's already created in #720. For instance, jenkins let's you define the base url. This can include a folder: http://server/folder. This way we can use it directly or setup a reverse proxy like http://server2/folder2 <-> http://server/folder.
I would gladly make valid contributions but coding this is really over my head.

**Reply by Radiergummi** [reply]
I think it isn't really that complicated actually! Looking at [the server package](https://github.com/lowdefy/lowdefy/blob/main/packages/server/src/index.js#L70-L90), the following code _should_ be enough to have a configurable URI prefix:
```js
  // ...
  const baseUrl = get(appConfig, 'server.baseUrl', { default: '' });

  gqlServer.applyMiddleware({
    app: server,
    path: gqlExpressPath || baseUrl + '/api/graphql',  // <-- baseUrl added here 
    bodyParserConfig: { limit: '5mb' },
  });

  if (serveStaticFiles) {
    // serve index.html with appended html
    // else static server serves without appended html
    server.get(baseUrl + '/', serveIndex);  // <-- baseUrl added here

    server.use(baseUrl + '/shell', express.static(path.resolve(shellDirectory)));  // <-- baseUrl added here

    // serve public files
    server.use(baseUrl + '/public', express.static(path.resolve(publicDirectory)));  // <-- baseUrl added here

    // Redirect all 404 to index.html with status 200
    // This should always be the last route
    server.use(serveIndex); // baseUrl *could* be added here, I think
  }
  return server;
```

**Comment by SamTolmay** [comment]
In version 3.20 we released support for specifying the `basePath` of the Lowdefy server. This can be done by setting the `LOWDEFY_SERVER_BASE_PATH` environment variable. 

You can read more [here](https://docs.lowdefy.com/docker)
